xtask/src/themekeyscheck.rs

```diff
@@ -0,0 +1,430 @@
+                            .or_default()
+                            .push(language_name.clone());
+                    }
+                }
+            }
+        }
+    }
+
+    let mut has_error = false;
+    for (hq, langs) in highlights_queries {
+        if !is_known_key(&hq) {
+            eprintln!(
+                "'{}' is not a static theme key (seen in languages: {:?}) ",
+                hq, langs
+            );
+            has_error = true;
+        }
+    }
+
+    if has_error {
+        Err("Theme key are incomplete, add missing keys to the list in `themekeyscheck.rs`".into())
+    } else {
+        println!("All theme keys found in highlight queries are known !");
+        Ok(())
+    }
+}
+
+/// Generate the doc table for `book/src/theme.md`
+pub fn theme_keys_doc() -> String {
+    fn max_size(lengths: &[(&'static str, &'static str)]) -> (usize, usize) {
+        lengths.iter().fold((0, 0), |(km, nm), (kl, nl)| {
+            (km.max(kl.len()), nm.max(nl.len()))
+        })
+    }
+
+    let (keys_w, notes_w) = max_size(KEY_DOCS);
+
+    let mut table = String::from(
+        "<!-- DO NOT MODIFY MANUALLY: Document generated by xtask::themekeyscheck::theme_keys_doc() using `cargo xtask docgen` -->",
+    );
+
+    let write_table_header = |table: &mut String, keys_type: &'static str| {
+        table.push('\n');
+        ::std::writeln!(
+            table,
+            "| {k:kw$} | {n:nw$} |",
+            k = keys_type,
+            kw = keys_w,
+            n = "Notes",
+            nw = notes_w
+        )
+        .unwrap();
+        ::std::writeln!(
+            table,
+            "| {k:kw$} | {n:nw$} |",
+            k = "---",
+            kw = keys_w,
+            n = "---",
+            nw = notes_w
+        )
+        .unwrap();
+    };
+
+    write_table_header(&mut table, "UI-only Keys");
+
+    for &(key, note) in KEY_DOCS {
+        if key == FIRST_MIXED_KEY {
+            write_table_header(&mut table, "Mixed Keys (UI & languages)");
+        } else if key == FIRST_LANGUAGE_KEY {
+            write_table_header(&mut table, "Language-only Keys");
+        }
+
+        ::std::writeln!(
+            &mut table,
+            "| {k:kw$} | {n:nw$} |",
+            k = key,
+            kw = keys_w,
+            n = note,
+            nw = notes_w
+        )
+        .unwrap();
+    }
+
+    table.push('\n');
+    table.push_str("[e-sec]: ./configuration.md#editor-section\n");
+    table.push_str("[e-sl-sec]: ./configuration.md#editorstatusline-section\n");
+    table.push_str("[e-ig-sec]: ./configuration.md#editorindent-guides-section\n");
+    table.push_str("[e-ws-sec]: ./configuration.md#editorwhitespace-section\n");
+
+    table
+}
+
+/// Keys must be given in the following order:
+///
+/// - UI keys (used only in UI theming)
+/// - Mixed keys (used both for language and UI theming)
+/// - Language keys (used only in language theming)
+macro_rules! generate_theme_keys {
```

Comment:
I think this would be more approachable to newer Rustaceans if it used the common `hash_map` macro (or `hashmap!` from helix_core) for associating the keys and docs. The other constants in here could just be regular constants defined outside of the hashmap macro

Reply:
I don't agree, it would create a big allocation for nothing in a check that is already quite slow and would not allow iterating in insertion order, which is important if we don't want this to create big diffs on each changes.